tsl/src/compression/compress_utils.c

```diff
@@ -573,6 +574,10 @@ bool
 tsl_recompress_chunk_wrapper(Chunk *uncompressed_chunk)
 {
 	Oid uncompressed_chunk_relid = uncompressed_chunk->table_id;
+	/* To avoid a deadlock, we conservatively lock the uncompressed chunk when
+	 * doing a recompress. This will prevent reads from the uncompressed chunk
+	 * while recompress chunk is running. */
+	LockRelationOid(uncompressed_chunk_relid, AccessExclusiveLock);
```

Comment:
What about unlocking after recompress is done? This should help with performance, but all the locks need to be released to avoid the deadlock.

Reply:
Relation is automatically unlocked when the transaction commits, which is when recompress is done.

Reply:
Can't recompress be done as part of a multi-step transaction block?

Reply:
Ok, I guess it always will take place at the end, regardless of the isolation mode, is what you're saying?

Reply:
Yes, but in that case the user has chosen to start a transaction and we should not commit it prematurely nor release the locks prematurely because we think it is a good idea. This can create consistency issues. The normal use case is to run it at top level, and in this case it would be released on completion of the statement.

Reply:
The problem with taking a lock early in the process is that anything that needs access to this chunk will be blocked. All queries on this hypertable or any operations on this hypertable are completely blocked and will  be waiting for this lock. Recompress could be a very long operation. The deadlock issue was due to the order in which the locks were acquired. Some of the locks can be completely eliminated by avoiding some steps during recompress as outlined in the support issue. Why are we not implementing that approach instead of taking more heavy handed locks?

Reply:
Removing the strong lock in favor of not re-building the index for the uncompressed chunk, as we discussed. This will avoid taking an `AccessExclusive` lock on the uncompressed chunk and eliminate the immediate problem. Other improvements are possible as well, but we should take these as a separate PR.